File: java/longest_substring_k_distinct.java
class Solution {
    public int lengthOfLongestSubstringKDistinct(String s, int k) {
        // if (s == null || s.length() == 0 || k == 0) return 0;
        HashMap<Character, Integer> d = new HashMap<Character, Integer>();
        int left = 0, result = 0;
        for (int i = 0; i < s.length(); i++) {
            d.put(s.charAt(i), i);
            if (d.size() > k) {
                int min = Collections.min(d.values());
                left = min + 1;
                d.remove(s.charAt(min));
            }
            result = Math.max(result, i - left + 1);
        }
        return result;
    }
}

// If required to return all the substrings (or the # of substrings) to have exactly k distinct letters
class Solution {
    public String[] lengthOfLongestSubstringKDistinct(String s, int k) {
        // if (s == null || s.length() == 0 || k == 0) return 0;
        HashMap<Character, Integer> d = new HashMap<Character, Integer>();
        List<String> l = new ArrayList<String>();
        int left = 0, result = 0;
        for (int i = 0; i < s.length(); i++) {
            d.put(s.charAt(i), i);
            if (d.size == k) {
                l.add(s.substring(left, i + 1));
            }
            if (d.size() > k) {
                int min = Collections.min(d.values());
                left = min + 1;
                d.remove(s.charAt(min));
            }
            result = Math.max(result, i - left + 1);
        }
        return l.toArray(new String[l.size()]);
    }
}

File: python/longest_substring_k_distinct.py
def lengthOfLongestSubstringKDistinct(self, s, k):
    """
    :type s: str
    :type k: int
    :rtype: int
    """
    if not s or k == 0:
        return 0
    dic = collections.defaultdict(int)
    result = 0
    left = 0
    for i, letter in enumerate(s):
        dic[letter] += 1
        while len(dic) > k:
            dic[s[left]] -= 1
            if dic[s[left]] == 0:
                del dic[s[left]]
            left += 1
        result = max(result, i - left + 1)
    return result


# Or slightly better way:
def lengthOfLongestSubstringKDistinct(self, s, k):
    """
    :type s: str
    :type k: int
    :rtype: int
    """
    if not s or k == 0:
        return 0
    dic = {}
    left = 0
    result = 0
    for i, letter in enumerate(s):
        dic[letter] = i
        if len(dic) > k:
            to_del = min(dic.values())
            del dic[s[to_del]]
            left = to_del + 1
        result = max(result, i - left + 1)
    return result
